### PR TITLE
[TA-3481]: Add ownable interface to erc20 precompile

### PR DESCRIPTION
**Pull request by GuillemGarciaDev**
# [TA-3481]: Add ownable interface to erc20 precompile

# Description

This PR aims to implement the `Ownable` openzeppelin contract interface for erc20 precompile.
It includes a migration to migrate the `TokenPair` type with the new field `ContractOwnerAddress`.

---

## Author Checklist

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.

I have...

- [ ] tackled an existing issue or discussed with a team member
- [ ] left instructions on how to review the changes
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/evmos/evmos/blob/main/CONTRIBUTING.md#pr-targeting))

## Reviewers Checklist

**All** items are required.
Please add a note if the item is not applicable
and please add your handle next to the items reviewed
if you only reviewed selected items.

I have...

- [ ] added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] confirmed all author checklist items have been addressed
- [ ] confirmed that this PR does not change production code
- [ ] reviewed content
- [ ] tested instructions (if applicable)
- [ ] confirmed all CI checks have passed

### CODE REVIEW

File: proto/evmos/erc20/v1/erc20.proto [hunk 1]
```diff
@@ -30,6 +30,9 @@ message TokenPair {
   bool enabled = 3;
   // contract_owner is the an ENUM specifying the type of ERC20 owner (0 invalid, 1 ModuleAccount, 2 external address)
   Owner contract_owner = 4;
+  // contract_owner_address is the address of the current owner of the token
+  // Only set if contract_owner is OWNER_MODULE
+  string contract_owner_address = 5;
```
Comment (AdriaCarrera) [inline proto/evmos/erc20/v1/erc20.proto]: Això hauria de ser nullable

File: app/upgrades/v19/upgrades.go [hunk 2]
```diff
@@ -120,7 +120,7 @@ func RunSTRv2Migration(
 ) error {
 	// NOTE: it's necessary to register the WEVMOS token as a native token pair before adding
 	// the dynamic EVM extensions (which is relying on the registered token pairs).
-	pair := erc20types.NewTokenPair(wrappedContractAddr, nativeDenom, erc20types.OWNER_MODULE)
+	pair := erc20types.NewTokenPair(wrappedContractAddr, nativeDenom, erc20types.OWNER_MODULE, "")
```
Comment (AdriaCarrera) [inline app/upgrades/v19/upgrades.go]: Aquesta funció no s'hauria de modificar en cap cas ja que la upgrade ja està feta

File: x/erc20/types/token_pair.go [hunk 3]
```diff
@@ -30,12 +30,13 @@ func NewTokenPairSTRv2(denom string) (TokenPair, error) {
 }
 
 // NewTokenPair returns an instance of TokenPair
-func NewTokenPair(erc20Address common.Address, denom string, contractOwner Owner) TokenPair {
+func NewTokenPair(erc20Address common.Address, denom string, contractOwner Owner, contractOwnerAddr string) TokenPair {
```
Comment (AdriaCarrera) [inline x/erc20/types/token_pair.go]: Si posem contractOwnerAdddress com a nil, podem evitar canviar la firma d'aquesta funció i modificar tants fitxers

File: proto/evmos/erc20/v1/erc20.proto (hunk 1 above)
Comment (AdriaCarrera) [reply proto/evmos/erc20/v1/erc20.proto]: Cambiar el nom a `owner_address`

File: precompiles/erc20/erc20.go [hunk 4]
```diff
@@ -35,6 +35,10 @@ const (
 	GasTotalSupply       = 2_477
 	GasBalanceOf         = 2_851
 	GasAllowance         = 3_246
+	// TODO: Set gas for MintMethod, BurnMethod, and Ownable methods
+	GasMint              = 3_000_000
```
Comment (AdriaCarrera) [inline precompiles/erc20/erc20.go]: Comprovar quins son els valors reals

File: precompiles/erc20/erc20.go [hunk 5]
```diff
@@ -47,27 +51,30 @@ var _ vm.PrecompiledContract = &Precompile{}
 // Precompile defines the precompiled contract for ERC-20.
 type Precompile struct {
 	cmn.Precompile
+	storeKey       storetypes.StoreKey
```
Comment (AdriaCarrera) [inline precompiles/erc20/erc20.go]: Posar el Erc20Keeper aquí

File: precompiles/erc20/tx.go [hunk 6]
```diff
@@ -239,4 +254,49 @@ func (p *Precompile) Burn(
...
+
+// TransferOwnership executes a transfer of ownership of the token.
+func (p *Precompile) TransferOwnership(
+	ctx sdk.Context,
+	contract *vm.Contract,
+	stateDB vm.StateDB,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	newOwner, err := ParseTransferOwnershipArgs(args)
```
Comment (AdriaCarrera) [inline precompiles/erc20/tx.go:234]: Moure la llogica al Erc20Keeper

**Review by AdriaCarrera — CHANGES_REQUESTED** [review]

File: precompiles/erc20/errors.go [hunk 7]
```diff
@@ -82,6 +86,10 @@ func ConvertErrToERC20Error(err error) error {
 		return ErrDecreasedAllowanceBelowZero
 	case strings.Contains(err.Error(), cmn.ErrIntegerOverflow):
 		return vm.ErrExecutionReverted
+	case strings.Contains(err.Error(), "ERC20: unauthorized") || strings.Contains(err.Error(), "authorization not found"):
```
Comment (AdriaCarrera) [inline precompiles/erc20/errors.go]: Per que aquest tenen el prefix `ERC20:` ?

File: precompiles/erc20/interfaces.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type Keeper interface {
+	// Mint mints new tokens to the given address
+	MintCoins(ctx sdk.Context, sender, to sdk.AccAddress, amount math.Int, token string) error
+	// Burn burns tokens from the given address
+	BurnCoins(ctx sdk.Context, sender sdk.AccAddress, amount math.Int, token string) error
+	// OwnerAddress returns the owner address of the token
+	GetTokenPairOwnerAddress(ctx sdk.Context, token string) (sdk.AccAddress, error)
```
Comment (AdriaCarrera) [inline precompiles/erc20/interfaces.go:14]: No seria millor passar-li tot el `TokenPair` enlloc de un string?

File: precompiles/erc20/interfaces.go [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+	// TransferOwnership transfers ownership of the token to the new owner
+	TransferOwnership(ctx sdk.Context, sender, newOwner sdk.AccAddress, token string) error
```
Comment (AdriaCarrera) [inline precompiles/erc20/interfaces.go]: Ídem

File: precompiles/erc20/query.go [hunk 10]
```diff
@@ -206,6 +209,29 @@ func (p Precompile) Allowance(
...
+	if err != nil {
+		return nil, err
+	}
+
+	ownerAddr, err := p.erc20Keeper.GetTokenPairOwnerAddress(ctx, p.tokenPair.GetERC20Contract().Hex())
+	if err != nil {
+		return nil, err
+	}
+
+	p.tokenPair.SetOwnerAddress(ownerAddr.String())
```
Comment (AdriaCarrera) [inline precompiles/erc20/query.go]: Per que es fa un `SetOwnerAddress` aquí?

File: precompiles/erc20/tx.go [hunk 11]
```diff
@@ -158,25 +160,15 @@ func (p *Precompile) Mint(
...
+	toAddr := sdk.AccAddress(to.Bytes())
 
 	coins := sdk.Coins{{Denom: p.tokenPair.Denom, Amount: math.NewIntFromBigInt(amount)}}
-	err = p.bankKeeper.MintCoins(ctx, erc20types.ModuleName, coins)
-	if err != nil {
-		return nil, err
-	}
 
-	err = p.bankKeeper.SendCoinsFromModuleToAccount(ctx, erc20types.ModuleName, sdk.AccAddress(to.Bytes()), coins)
+	err = p.erc20Keeper.MintCoins(ctx, minter, toAddr, math.NewIntFromBigInt(amount), p.tokenPair.GetERC20Contract().Hex())
```
Comment (AdriaCarrera) [inline precompiles/erc20/tx.go:174]: No sería millor passar-li tot el `TokenPair` enlloc de `GetERC20Contract`?

File: precompiles/erc20/tx.go [hunk 12]
```diff
@@ -208,35 +200,51 @@ func (p *Precompile) Burn(
...
-	// burnerIsOwner := burner.Equals(sdk.AccAddress(from.Bytes()))
-	burnerIsOwner := true
 
-	if !burnerIsOwner {
-		return nil, ConvertErrToERC20Error(errorsmod.Wrapf(authz.ErrNoAuthorizationFound, "burner is not the owner"))
-	}
 	coins := sdk.Coins{{Denom: p.tokenPair.Denom, Amount: math.NewIntFromBigInt(amount)}}
 
-	err = p.bankKeeper.SendCoinsFromAccountToModule(ctx, burner, erc20types.ModuleName, coins)
+	err = p.erc20Keeper.BurnCoins(ctx, burner, math.NewIntFromBigInt(amount), p.tokenPair.GetERC20Contract().Hex())
```
Comment (AdriaCarrera) [inline precompiles/erc20/tx.go:209]: Ídem

File: precompiles/erc20/tx.go [hunk 13]
```diff
@@ -208,35 +200,51 @@ func (p *Precompile) Burn(
...
 	if err != nil {
 		return nil, err
 	}
 
-	if p.tokenPair.Denom == utils.BaseDenom {
-		p.SetBalanceChangeEntries(
-			cmn.NewBalanceChangeEntry(burnerAddr, coins.AmountOf(utils.BaseDenom).BigInt(), cmn.Sub))
+	sender := sdk.AccAddress(contract.CallerAddress.Bytes())
+
+	err = p.erc20Keeper.TransferOwnership(ctx, sender, sdk.AccAddress(newOwner.Bytes()), p.tokenPair.GetERC20Contract().Hex())
```
Comment (AdriaCarrera) [inline precompiles/erc20/tx.go]: Ídem

File: precompiles/erc20/tx_test.go [hunk 14]
```diff
@@ -265,47 +266,82 @@ func (s *PrecompileTestSuite) TestMint() {
...
+				tokenPair := erc20types.NewTokenPair(utiltx.GenerateAddress(), s.tokenDenom, erc20types.OWNER_MODULE)
+				tokenPair.SetOwnerAddress(sdk.AccAddress(utiltx.GenerateAddress().Bytes()).String())
+				s.network.App.Erc20Keeper.SetTokenPair(s.network.GetContext(), tokenPair)
+				s.network.App.Erc20Keeper.SetDenomMap(s.network.GetContext(), tokenPair.Denom, tokenPair.GetID())
+				s.network.App.Erc20Keeper.SetERC20Map(s.network.GetContext(), tokenPair.GetERC20Contract(), tokenPair.GetID())
+				return []interface{}{spender.Addr, big.NewInt(100)}, tokenPair
+			},
+			func() {},
+			true,
+			"execution reverted",
```
Comment (AdriaCarrera) [inline precompiles/erc20/tx_test.go]: Aquest error es poc descriptiu, no?

File: precompiles/erc20/tx_test.go [hunk 15]
```diff
@@ -403,3 +445,75 @@ func (s *PrecompileTestSuite) TestBurn() {
...
+			expErr:      true,
+			errContains: "invalid new owner address",
+		},
+		{
+			name: "pass",
+			malleate: func() []interface{} {
+				return []interface{}{newOwner}
+			},
+			postCheck: func(precompile *erc20.Precompile) {
+				// owner, err := precompile.GetContractOwnerAddress(s.network.GetContext())
```
Comment (AdriaCarrera) [inline precompiles/erc20/tx_test.go]: Borrar comments

File: proto/evmos/erc20/v1/erc20.proto [hunk 16]
```diff
@@ -30,6 +30,9 @@ message TokenPair {
   bool enabled = 3;
   // contract_owner is the an ENUM specifying the type of ERC20 owner (0 invalid, 1 ModuleAccount, 2 external address)
   Owner contract_owner = 4;
+  // contract_owner_address is the address of the current owner of the token
```
Comment (AdriaCarrera) [inline proto/evmos/erc20/v1/erc20.proto]: Fix comment `contract_owner_address` és el nom antic

File: proto/evmos/erc20/v1/tx.proto [hunk 17]
```diff
@@ -71,3 +80,29 @@ message MsgUpdateParams {
 // MsgUpdateParams message.
 // Since: cosmos-sdk 0.47
 message MsgUpdateParamsResponse {}
+
+// MsgMint defines a Msg to mint ERC20 tokens
+message MsgMint {
+  // contract_address of an ERC20 token contract, that is registered in a token pair
+  string contract_address = 1;
```
Comment (AdriaCarrera) [inline proto/evmos/erc20/v1/tx.proto:111]: Aquest no hauria de tenir el type address? `(cosmos_proto.scalar) = "cosmos.AddressString"`

File: proto/evmos/erc20/v1/tx.proto [hunk 18]
```diff
@@ -71,3 +80,29 @@ message MsgUpdateParams {
...
+  // amount of ERC20 tokens to mint
+  string amount = 2 [(gogoproto.customtype) = "cosmossdk.io/math.Int", (gogoproto.nullable) = false];
+  // to is the address to mint the tokens to
```
Comment (AdriaCarrera) [inline proto/evmos/erc20/v1/tx.proto:114]: Ídem

File: proto/evmos/erc20/v1/tx.proto [hunk 19]
```diff
@@ -71,3 +80,29 @@ message MsgUpdateParams {
...
+  // sender is the address of the sender
+  string sender = 4;
+}
+
+message MsgMintResponse {}
+
+// MsgBurn defines a Msg to burn ERC20 tokens
+message MsgBurn {
+  // contract_address of an ERC20 token contract, that is registered in a token pair
+  string contract_address = 1;
```
Comment (AdriaCarrera) [inline proto/evmos/erc20/v1/tx.proto:125]: Ídem

File: proto/evmos/erc20/v1/tx.proto [hunk 20]
```diff
@@ -71,3 +80,29 @@ message MsgUpdateParams {
...
+  // amount of ERC20 tokens to burn
+  string amount = 2 [(gogoproto.customtype) = "cosmossdk.io/math.Int", (gogoproto.nullable) = false];
+  // sender is the address of the sender
+  string sender = 3;
```
Comment (AdriaCarrera) [inline proto/evmos/erc20/v1/tx.proto:129]: Ídem

File: x/erc20/keeper/erc20.go [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+		)
+	}
+
+	if !pair.Enabled {
+		return types.TokenPair{}, errorsmod.Wrapf(
+			types.ErrERC20TokenPairDisabled, "minting token '%s' is not enabled by governance", token,
+		)
+	}
+
+	if k.bankKeeper.BlockedAddr(receiver.Bytes()) {
```
Comment (AdriaCarrera) [inline x/erc20/keeper/erc20.go:52]: I aquesta assertion? Aplica també al transfer?

File: x/erc20/keeper/erc20.go [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = k.bankKeeper.BurnCoins(ctx, types.ModuleName, coins)
+	if err != nil {
+		return err
+	}
+
+	// if p.tokenPair.Denom == utils.BaseDenom {
```
Comment (AdriaCarrera) [inline x/erc20/keeper/erc20.go]: Borrar comments

File: x/erc20/keeper/token_pairs.go [hunk 23]
```diff
@@ -203,3 +203,29 @@ func (k Keeper) GetTokenDenom(ctx sdk.Context, tokenAddress common.Address) (str
 
 	return tokenPair.Denom, nil
 }
+
+// GetTokenOwnerAddress returns the OwnerAddress of the token
+func (k Keeper) GetTokenPairOwnerAddress(ctx sdk.Context, token string) (sdk.AccAddress, error) {
+	tokenPairID := k.GetERC20Map(ctx, common.HexToAddress(token))
```
Comment (AdriaCarrera) [inline x/erc20/keeper/token_pairs.go:209]: No és pot fer més eficient aquesta busqueda?

File: x/erc20/keeper/token_pairs.go [hunk 24]
```diff
@@ -203,3 +203,29 @@ func (k Keeper) GetTokenDenom(ctx sdk.Context, tokenAddress common.Address) (str
...
+	return sdk.AccAddressFromBech32(tokenPair.OwnerAddress)
+}
+
+// SetTokenPairOwnerAddress sets the owner address for the token pair
+func (k Keeper) SetTokenPairOwnerAddress(ctx sdk.Context, pair types.TokenPair, newOwner string) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixTokenPair)
+
+	pair.SetOwnerAddress(newOwner)
+
+	key := pair.GetID()
```
Comment (AdriaCarrera) [inline x/erc20/keeper/token_pairs.go]: No es pot aprofitar el `SetTokenPair` ?

File: x/erc20/types/token_pair_test.go [hunk 25]
```diff
@@ -57,10 +57,10 @@ func (suite *TokenPairTestSuite) TestTokenPair() {
 		pair       types.TokenPair
 		expectPass bool
 	}{
-		{msg: "Register token pair - invalid address (no hex)", pair: types.TokenPair{"0x5dCA2483280D9727c80b5518faC4556617fb19ZZ", "test", true, types.OWNER_MODULE}, expectPass: false},
-		{msg: "Register token pair - invalid address (invalid length 1)", pair: types.TokenPair{"0x5dCA2483280D9727c80b5518faC4556617fb19", "test", true, types.OWNER_MODULE}, expectPass: false},
-		{msg: "Register token pair - invalid address (invalid length 2)", pair: types.TokenPair{"0x5dCA2483280D9727c80b5518faC4556617fb194FFF", "test", true, types.OWNER_MODULE}, expectPass: false},
-		{msg: "pass", pair: types.TokenPair{utiltx.GenerateAddress().String(), "test", true, types.OWNER_MODULE}, expectPass: true},
+		{msg: "Register token pair - invalid address (no hex)", pair: types.TokenPair{"0x5dCA2483280D9727c80b5518faC4556617fb19ZZ", "test", true, types.OWNER_MODULE, ""}, expectPass: false},
```
Comment (AdriaCarrera) [inline x/erc20/types/token_pair_test.go:60]: Treure el `""` ja que per default es queda buit

**Review by AdriaCarrera — CHANGES_REQUESTED** [review]

File: x/erc20/types/token_pair_test.go (hunk 25 above)
Comment (GuillemGarciaDev) [reply x/erc20/types/token_pair_test.go:60]: Si s'inicialitza un struct sense les labels dels fields s'han d'especificar tots els camps

**Pull request merged**